File: client/platform/android/src/main/kotlin/com/protectedvm/host/PvmModuleStore.kt
package com.protectedvm.host

import android.system.Os
import android.util.Base64
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.io.FileOutputStream
import java.net.HttpURLConnection
import java.net.URI
import java.security.MessageDigest

class PvmModuleStore(
    private val root: File,
    private val publicKeyPath: String,
    private val applicationId: String,
    private val channel: String,
    private val profile: String,
    serverBase: String,
    private val activationToken: String? = null,
    private val installationId: String? = null,
    allowHttpLocalhost: Boolean = false,
    private val minimumRelease: Long = 0,
) {
    private val server = URI(serverBase.trimEnd('/'))
    private val modules = File(root, "modules")
    private val currentFile = File(root, "current.json")

    init {
        require(
            server.scheme == "https" ||
                (
                    allowHttpLocalhost &&
                        server.scheme == "http" &&
                        server.host in setOf("127.0.0.1", "localhost")
                    )
        ) { "Module service must use HTTPS" }
        require(
            SEGMENT.matches(applicationId) &&
                SEGMENT.matches(channel) &&
                applicationId !in setOf(".", "..") &&
                channel !in setOf(".", "..")
        )
        require(profile in PROFILES)
        require(minimumRelease >= 0)
        modules.mkdirs()
    }

    fun lastKnownGood(): File? {
        val state = readState() ?: return null
        if (state.release < minimumRelease) return null
        return File(modules, "${state.sha256}.pvm").takeIf {
            it.isFile && it.length() in 1..MAX_MODULE_BYTES && it.hasSha256(state.sha256)
        }?.also {
            runCatching { Os.chmod(it.absolutePath, 0b110000000) }
        }
    }

    /**
     * Runs on a worker thread. A refresh failure never removes or replaces last-known-good.
     */
    fun refresh(): File {
        check(android.os.Looper.myLooper() != android.os.Looper.getMainLooper()) {
            "Module refresh must not block the UI thread"
        }
        val previous = readState()
        return runCatching { download(previous) }.getOrElse { error ->
            lastKnownGood() ?: throw error
        }
    }

    private fun download(previous: State?): File {
        val manifestPath = "/v1/apps/$applicationId/$channel/android/$profile/manifest"
        val connection = open(server.resolve(manifestPath))
        val cached = lastKnownGood()
        previous
            ?.takeIf { cached != null && it.release >= minimumRelease }
            ?.etag
            ?.takeIf(String::isNotEmpty)
            ?.let {
            connection.setRequestProperty("If-None-Match", it)
        }
        installationId?.let { connection.setRequestProperty("X-PVM-Installation-ID", it) }
        connection.connect()
        if (connection.responseCode == HttpURLConnection.HTTP_NOT_MODIFIED) {
            return requireNotNull(lastKnownGood()) { "Server returned 304 without a cached module" }
        }
        require(connection.responseCode == HttpURLConnection.HTTP_OK) {
            "Manifest request failed with HTTP ${connection.responseCode}"
        }
        val manifestBytes = connection.inputStream.use { it.readBounded(MAX_MANIFEST_BYTES) }
        val envelope = JSONObject(manifestBytes.toString(Charsets.UTF_8))
        require(envelope.getInt("envelope_format") == 1)
        require(envelope.getString("signature_algorithm") == "Ed25519")
        val payload = Base64.decode(envelope.getString("payload"), Base64.NO_WRAP)
        val signature = Base64.decode(envelope.getString("signature"), Base64.NO_WRAP)
        require(signature.size == 64 && PvmCrypto.verify(publicKeyPath, payload, signature)) {
            "Manifest signature verification failed"
        }
        val manifest = JSONObject(payload.toString(Charsets.UTF_8))
        require(manifest.getString("application_id") == applicationId)
        require(manifest.getString("channel") == channel)
        require(manifest.getString("profile") == profile)
        require(manifest.getString("platform") == "android")
        val release = manifest.getLong("release")
        val releaseFloor = maxOf(previous?.release ?: 0, minimumRelease)
        require(release >= releaseFloor) { "Manifest rejected by anti-rollback policy" }
        val digest = manifest.getString("sha256")
        require(SHA256.matches(digest)) { "Manifest contains an invalid SHA-256" }
        val size = manifest.getLong("size")
        require(size in 1..MAX_MODULE_BYTES) { "Module size is outside the host budget" }
        val destination = File(modules, "$digest.pvm")
        if (
            destination.isFile &&
            (destination.length() != size || !destination.hasSha256(digest))
        ) {
            require(destination.delete()) { "Cannot remove corrupt cached module" }
        }

        if (!destination.isFile) {
            val temporary = File(modules, "$digest.tmp")
            try {
                val moduleUri = server.resolve(manifest.getString("module_url"))
                require(
                    moduleUri.host == server.host &&
                        moduleUri.scheme == server.scheme &&
                        moduleUri.port == server.port &&
                        moduleUri.path == "/v1/modules/$digest.pvm"
                ) {
                    "Manifest module URL changed origin or hash binding"
                }
                val moduleConnection = open(moduleUri)
                moduleConnection.connect()
                require(moduleConnection.responseCode == HttpURLConnection.HTTP_OK) {
                    "Module request failed with HTTP ${moduleConnection.responseCode}"
                }
                val hash = MessageDigest.getInstance("SHA-256")
                var written = 0L
                FileOutputStream(temporary).use { output ->
                    moduleConnection.inputStream.use { input ->
                        val buffer = ByteArray(16_384)
                        while (true) {
                            val count = input.read(buffer)
                            if (count < 0) break
                            written += count
                            require(written <= size) { "Module exceeds its declared size" }
                            hash.update(buffer, 0, count)
                            output.write(buffer, 0, count)
                        }
                    }
                    output.fd.sync()
                }
                require(written == size) { "Module download is truncated" }
                require(hash.digest().toHex() == digest) { "Module SHA-256 mismatch" }
                val validatedRelease =
                    PvmModuleValidator.validate(
                        temporary.absolutePath,
                        publicKeyPath,
                        applicationId,
                        channel,
                        profile,
                        releaseFloor,
                    )
                require(validatedRelease == release) { "Manifest/module release mismatch" }
                Os.rename(temporary.absolutePath, destination.absolutePath)
                Os.chmod(destination.absolutePath, 0b110000000)
            } finally {
                temporary.delete()
            }
        } else {
            val validatedRelease =
                PvmModuleValidator.validate(
                    destination.absolutePath,
                    publicKeyPath,
                    applicationId,
                    channel,
                    profile,
                    releaseFloor,
                )
            require(validatedRelease == release) { "Cached manifest/module release mismatch" }
        }

        val history =
            buildList {
                add(digest)
                previous?.history.orEmpty().filterTo(this) {
                    it != digest && File(modules, "$it.pvm").isFile
                }
            }.distinct().take(2)
        writeState(
            State(
                etag = connection.getHeaderField("ETag").orEmpty(),
                release = release,
                sha256 = digest,
                history = history,
            )
        )
        modules.listFiles { file -> file.extension == "pvm" }?.forEach {
            if (it.nameWithoutExtension !in history) it.delete()
        }
        return destination
    }

    private fun open(uri: URI): HttpURLConnection =
        (uri.toURL().openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            instanceFollowRedirects = false
            connectTimeout = 10_000
            readTimeout = 20_000
            activationToken?.let { setRequestProperty("Authorization", "Bearer $it") }
        }

    private fun readState(): State? =
        runCatching {
            require(currentFile.length() in 1..MAX_STATE_BYTES)
            val source = JSONObject(currentFile.readText())
            require(source.getInt("format") == 1)
            require(source.getString("application_id") == applicationId)
            require(source.getString("channel") == channel)
            require(source.getString("platform") == "android")
            require(source.getString("profile") == profile)
            val digest = source.getString("sha256")
            require(SHA256.matches(digest))
            val historyArray = source.getJSONArray("history")
            val history =
                buildList {
                    for (index in 0 until historyArray.length()) {
                        val value = historyArray.getString(index)
                        require(SHA256.matches(value))
                        add(value)
                    }
                }
            require(history.isNotEmpty() && history.size <= 2)
            require(history.first() == digest && history.distinct().size == history.size)
            val release = source.getLong("release")
            require(release > 0)
            State(
                etag = source.optString("etag"),
                release = release,
                sha256 = digest,
                history = history,
            )
        }.getOrNull()

    private fun writeState(state: State) {
        root.mkdirs()
        val temporary = File(root, "current.tmp")
        val json =
            JSONObject()
                .put("format", 1)
                .put("application_id", applicationId)
                .put("channel", channel)
                .put("platform", "android")
                .put("profile", profile)
                .put("etag", state.etag)
                .put("release", state.release)
                .put("sha256", state.sha256)
                .put("history", JSONArray(state.history))
                .toString()
        FileOutputStream(temporary).use {
            it.write(json.toByteArray())
            it.fd.sync()
        }
        Os.rename(temporary.absolutePath, currentFile.absolutePath)
        Os.chmod(currentFile.absolutePath, 0b110000000)
    }

    private data class State(
        val etag: String,
        val release: Long,
        val sha256: String,
        val history: List<String>,
    )

    companion object {
        private val SEGMENT = Regex("[A-Za-z0-9._-]{1,255}")
        private val SHA256 = Regex("[0-9a-f]{64}")
        private val PROFILES =
            setOf(
                "offline_sealed",
                "online_provisioned",
                "store_on_demand",
                "enterprise_managed",
            )
        private const val MAX_MANIFEST_BYTES = 64 * 1024
        private const val MAX_MODULE_BYTES = 16L * 1024L * 1024L
        private const val MAX_STATE_BYTES = 16L * 1024L
    }
}

private fun java.io.InputStream.readBounded(maximum: Int): ByteArray {
    val result = java.io.ByteArrayOutputStream()
    val buffer = ByteArray(8_192)
    while (true) {
        val count = read(buffer)
        if (count < 0) break
        require(result.size() + count <= maximum) { "Response exceeds its size budget" }
        result.write(buffer, 0, count)
    }
    return result.toByteArray()
}

private fun ByteArray.toHex(): String = joinToString("") { "%02x".format(it) }

private fun File.hasSha256(expected: String): Boolean =
    runCatching {
        val hash = MessageDigest.getInstance("SHA-256")
        inputStream().use { input ->
            val buffer = ByteArray(16_384)
            while (true) {
                val count = input.read(buffer)
                if (count < 0) break
                hash.update(buffer, 0, count)
            }
        }
        hash.digest().toHex() == expected
    }.getOrDefault(false)

File: client/platform/ios/swift/PVMModuleStore.swift
import CryptoKit
import Foundation

public actor PVMModuleStore {
    public struct Configuration: Sendable {
        public let root: URL
        public let publicKeyPath: String
        public let applicationId: String
        public let channel: String
        public let profile: String
        public let server: URL
        public let activationToken: String?
        public let installationId: String?
        public let allowHTTPForLocalhost: Bool
        public let minimumRelease: UInt64

        public init(
            root: URL,
            publicKeyPath: String,
            applicationId: String,
            channel: String,
            profile: String,
            server: URL,
            activationToken: String? = nil,
            installationId: String? = nil,
            allowHTTPForLocalhost: Bool = false,
            minimumRelease: UInt64 = 0
        ) {
            self.root = root
            self.publicKeyPath = publicKeyPath
            self.applicationId = applicationId
            self.channel = channel
            self.profile = profile
            self.server = server
            self.activationToken = activationToken
            self.installationId = installationId
            self.allowHTTPForLocalhost = allowHTTPForLocalhost
            self.minimumRelease = minimumRelease
        }
    }

    public typealias Validator = @Sendable (_ module: URL, _ minimumRelease: UInt64) async throws -> UInt64

    private let config: Configuration
    private let validate: Validator
    private let fileManager = FileManager.default
    private let redirectRejector = PVMRedirectRejector()

    public init(configuration: Configuration, validator: @escaping Validator) throws {
        guard Self.isSegment(configuration.applicationId),
              Self.isSegment(configuration.channel),
              configuration.applicationId != ".", configuration.applicationId != "..",
              configuration.channel != ".", configuration.channel != "..",
              Self.profiles.contains(configuration.profile)
        else { throw PVMHostError("Invalid module store binding") }
        let localHTTP =
            configuration.allowHTTPForLocalhost &&
            configuration.server.scheme == "http" &&
            ["localhost", "127.0.0.1"].contains(configuration.server.host ?? "")
        guard configuration.server.scheme == "https" || localHTTP else {
            throw PVMHostError("Module service must use HTTPS")
        }
        self.config = configuration
        self.validate = validator
    }

    public func lastKnownGood() -> URL? {
        guard let state = readState() else { return nil }
        guard state.release >= config.minimumRelease else { return nil }
        let module = modules.appendingPathComponent("\(state.sha256).pvm")
        guard fileManager.fileExists(atPath: module.path),
              let size = fileSize(module),
              (1...Self.maximumModuleBytes).contains(size),
              (try? sha256(module)) == state.sha256
        else { return nil }
        try? fileManager.setAttributes(
            [.protectionKey: FileProtectionType.completeUntilFirstUserAuthentication],
            ofItemAtPath: module.path
        )
        return module
    }

    public func refresh(session: URLSession = .shared) async throws -> URL {
        do {
            return try await refreshStrict(session: session)
        } catch {
            if let cached = lastKnownGood() { return cached }
            throw error
        }
    }

    private func refreshStrict(session: URLSession) async throws -> URL {
        try fileManager.createDirectory(at: modules, withIntermediateDirectories: true)
        let previous = readState()
        let manifestURL =
            config.server
                .appendingPathComponent("v1")
                .appendingPathComponent("apps")
                .appendingPathComponent(config.applicationId)
                .appendingPathComponent(config.channel)
                .appendingPathComponent("ios")
                .appendingPathComponent(config.profile)
                .appendingPathComponent("manifest")
        var request = URLRequest(url: manifestURL, timeoutInterval: 15)
        request.cachePolicy = .reloadIgnoringLocalCacheData
        if let token = config.activationToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        let cached = lastKnownGood()
        if let previous, cached != nil,
           previous.release >= config.minimumRelease, !previous.etag.isEmpty {
            let etag = previous.etag
            request.setValue(etag, forHTTPHeaderField: "If-None-Match")
        }
        if let installationId = config.installationId {
            request.setValue(installationId, forHTTPHeaderField: "X-PVM-Installation-ID")
        }
        let (manifestData, response) = try await boundedData(for: request, session: session, maximum: 64 * 1024)
        guard let http = response as? HTTPURLResponse else {
            throw PVMHostError("Manifest response is not HTTP")
        }
        guard http.url == manifestURL else {
            throw PVMHostError("Manifest request was redirected")
        }
        if http.statusCode == 304 {
            guard let cached = lastKnownGood() else {
                throw PVMHostError("Server returned 304 without a cached module")
            }
            return cached
        }
        guard http.statusCode == 200 else {
            throw PVMHostError("Manifest request failed with HTTP \(http.statusCode)")
        }
        let envelope = try JSONDecoder().decode(SignedEnvelope.self, from: manifestData)
        guard envelope.envelopeFormat == 1,
              envelope.signatureAlgorithm == "Ed25519",
              let payload = Data(base64Encoded: envelope.payload),
              let signature = Data(base64Encoded: envelope.signature),
              PVMPlatformCrypto.verify(
                payload: payload,
                signature: signature,
                publicKeyPath: config.publicKeyPath
              )
        else { throw PVMHostError("Manifest signature verification failed") }
        let manifest = try JSONDecoder().decode(Manifest.self, from: payload)
        let releaseFloor = max(previous?.release ?? 0, config.minimumRelease)
        guard manifest.applicationId == config.applicationId,
              manifest.channel == config.channel,
              manifest.profile == config.profile,
              manifest.platform == "ios",
              manifest.release >= releaseFloor,
              Self.isSHA256(manifest.sha256),
              (1...Self.maximumModuleBytes).contains(manifest.size)
        else { throw PVMHostError("Manifest binding, release, hash, or size is invalid") }

        let destination = modules.appendingPathComponent("\(manifest.sha256).pvm")
        if fileManager.fileExists(atPath: destination.path),
           fileSize(destination) != manifest.size ||
            (try? sha256(destination)) != manifest.sha256 {
            try fileManager.removeItem(at: destination)
        }
        if !fileManager.fileExists(atPath: destination.path) {
            guard let moduleURL = URL(string: manifest.moduleURL, relativeTo: config.server)?.absoluteURL,
                  moduleURL.scheme == config.server.scheme,
                  moduleURL.host == config.server.host,
                  moduleURL.port == config.server.port,
                  moduleURL.path == "/v1/modules/\(manifest.sha256).pvm"
            else { throw PVMHostError("Manifest module URL changed origin or hash binding") }
            var moduleRequest = URLRequest(url: moduleURL, timeoutInterval: 30)
            if let token = config.activationToken {
                moduleRequest.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            }
            let (moduleData, moduleResponse) =
                try await boundedData(
                    for: moduleRequest,
                    session: session,
                    maximum: manifest.size
                )
            guard let moduleHTTP = moduleResponse as? HTTPURLResponse,
                  moduleHTTP.statusCode == 200,
                  moduleHTTP.url == moduleURL,
                  moduleData.count == manifest.size
            else { throw PVMHostError("Module download was redirected, truncated, or rejected") }
            let temporary = modules.appendingPathComponent("\(manifest.sha256).tmp")
            try? fileManager.removeItem(at: temporary)
            try moduleData.write(to: temporary, options: .atomic)
            defer { try? fileManager.removeItem(at: temporary) }
            guard try sha256(temporary) == manifest.sha256
            else { throw PVMHostError("Module size or SHA-256 mismatch") }
            let release = try await validate(temporary, releaseFloor)
            guard release == manifest.release else {
                throw PVMHostError("Manifest/module release mismatch")
            }
            try fileManager.moveItem(at: temporary, to: destination)
            try fileManager.setAttributes(
                [.protectionKey: FileProtectionType.completeUntilFirstUserAuthentication],
                ofItemAtPath: destination.path
            )
        } else {
            let release = try await validate(destination, releaseFloor)
            guard release == manifest.release else {
                throw PVMHostError("Cached module release mismatch")
            }
        }

        let prior = previous?.history ?? []
        let history =
            ([manifest.sha256] + prior.filter {
                $0 != manifest.sha256 &&
                fileManager.fileExists(atPath: modules.appendingPathComponent("\($0).pvm").path)
            }).uniqued().prefix(2)
        let state =
            State(
                format: 1,
                applicationId: config.applicationId,
                channel: config.channel,
                platform: "ios",
                profile: config.profile,
                etag: http.value(forHTTPHeaderField: "ETag") ?? "",
                release: manifest.release,
                sha256: manifest.sha256,
                history: Array(history)
            )
        let encoded = try JSONEncoder().encode(state)
        try encoded.write(to: currentFile, options: [.atomic, .completeFileProtection])
        let keep = Set(state.history)
        for candidate in try fileManager.contentsOfDirectory(at: modules, includingPropertiesForKeys: nil)
        where candidate.pathExtension == "pvm" && !keep.contains(candidate.deletingPathExtension().lastPathComponent) {
            try? fileManager.removeItem(at: candidate)
        }
        return destination
    }

    private func readState() -> State? {
        guard let size = fileSize(currentFile),
              (1...Self.maximumStateBytes).contains(size),
              let data = try? Data(contentsOf: currentFile),
              let state = try? JSONDecoder().decode(State.self, from: data),
              state.format == 1,
              state.applicationId == config.applicationId,
              state.channel == config.channel,
              state.platform == "ios",
              state.profile == config.profile,
              state.release > 0,
              Self.isSHA256(state.sha256),
              !state.history.isEmpty,
              state.history.count <= 2,
              state.history.first == state.sha256,
              Set(state.history).count == state.history.count,
              state.history.allSatisfy(Self.isSHA256)
        else { return nil }
        return state
    }

    private func fileSize(_ file: URL) -> Int? {
        try? file.resourceValues(forKeys: [.fileSizeKey]).fileSize
    }

    private func boundedData(
        for request: URLRequest,
        session: URLSession,
        maximum: Int
    ) async throws -> (Data, URLResponse) {
        let (bytes, response) = try await session.bytes(
            for: request,
            delegate: redirectRejector
        )
        var data = Data()
        data.reserveCapacity(min(maximum, 16 * 1024))
        for try await byte in bytes {
            guard data.count < maximum else { throw PVMHostError("Response exceeds its size budget") }
            data.append(byte)
        }
        return (data, response)
    }

    private func sha256(_ file: URL) throws -> String {
        let handle = try FileHandle(forReadingFrom: file)
        defer { try? handle.close() }
        var hasher = SHA256()
        while true {
            let chunk = try handle.read(upToCount: 16 * 1024) ?? Data()
            if chunk.isEmpty { break }
            hasher.update(data: chunk)
        }
        return hasher.finalize().map { String(format: "%02x", $0) }.joined()
    }

    private var modules: URL { config.root.appendingPathComponent("modules", isDirectory: true) }
    private var currentFile: URL { config.root.appendingPathComponent("current.json") }

    private struct Manifest: Decodable {
        let applicationId: String
        let channel: String
        let profile: String
        let platform: String
        let release: UInt64
        let sha256: String
        let size: Int
        let moduleURL: String

        enum CodingKeys: String, CodingKey {
            case applicationId = "application_id"
            case channel, profile, platform, release, sha256, size
            case moduleURL = "module_url"
        }
    }

    private struct SignedEnvelope: Decodable {
        let envelopeFormat: Int
        let payload: String
        let signature: String
        let signatureAlgorithm: String

        enum CodingKeys: String, CodingKey {
            case envelopeFormat = "envelope_format"
            case payload, signature
            case signatureAlgorithm = "signature_algorithm"
        }
    }

    private struct State: Codable {
        let format: Int
        let applicationId: String
        let channel: String
        let platform: String
        let profile: String
        let etag: String
        let release: UInt64
        let sha256: String
        let history: [String]

        enum CodingKeys: String, CodingKey {
            case format
            case applicationId = "application_id"
            case channel, platform, profile, etag, release, sha256, history
        }
    }

    private static let profiles = [
        "offline_sealed",
        "online_provisioned",
        "store_on_demand",
        "enterprise_managed",
    ]
    private static let maximumModuleBytes = 16 * 1024 * 1024
    private static let maximumStateBytes = 16 * 1024

    private static func isSegment(_ value: String) -> Bool {
        let bytes = value.utf8
        return !bytes.isEmpty && bytes.count <= 255 &&
            bytes.allSatisfy {
                ($0 >= 48 && $0 <= 57) || ($0 >= 65 && $0 <= 90) ||
                ($0 >= 97 && $0 <= 122) || $0 == 46 || $0 == 95 || $0 == 45
            }
    }

    private static func isSHA256(_ value: String) -> Bool {
        let bytes = value.utf8
        return bytes.count == 64 &&
            bytes.allSatisfy { ($0 >= 48 && $0 <= 57) || ($0 >= 97 && $0 <= 102) }
    }
}

private extension Sequence where Element: Hashable {
    func uniqued() -> [Element] {
        var seen = Set<Element>()
        return filter { seen.insert($0).inserted }
    }
}

private final class PVMRedirectRejector: NSObject, URLSessionTaskDelegate, @unchecked Sendable {
    func urlSession(
        _ session: URLSession,
        task: URLSessionTask,
        willPerformHTTPRedirection response: HTTPURLResponse,
        newRequest request: URLRequest,
        completionHandler: @escaping (URLRequest?) -> Void
    ) {
        completionHandler(nil)
    }
}
